# L1: Information retrieval

In this lab you will apply basic techniques from information retrieval to implement the core of a minimalistic search engine. The data for this lab consists of a collection of app descriptions scraped from the [Google Play Store](https://play.google.com/store/apps?hl=en). From this collection, your search engine should retrieve those apps whose descriptions best match a given query under the vector space model.

**⚠️ Make sure that you have read the [Rules for hand-in assignments](https://www.ida.liu.se/~TDDE16/exam.en.shtml#handins) and the [Policy on cheating and plagiarism](https://www.ida.liu.se/~TDDE16/exam.en.shtml#cheating) before starting with this lab.** 

## Data set

The app descriptions come in the form of a compressed [JSON](https://en.wikipedia.org/wiki/JSON) file. Start by loading this file into a [Pandas](https://pandas.pydata.org) [DataFrame](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe).

In [3]:

import bz2
import pandas as pd
import spacy

with bz2.open('app-descriptions.json.bz2') as source:
    df = pd.read_json(source)

nlp = spacy.load("en_core_web_sm")

In Pandas, a DataFrame is a table with indexed rows and labelled columns of potentially different types. You can access data in a DataFrame in various ways, including by row and column. To give an example, the code in the next cell shows rows 200–204:

In [4]:
df[200:205]

name  \
200  Brick Breaker Star: Space King   
201      Brick Classic - Brick Game   
202     Bricks Breaker - Glow Balls   
203            Bricks Breaker Quest   
204             Brothers in Arms® 3   

                                           description  
200  Introducing the best Brick Breaker game that e...  
201  Classic Brick Game!\n\nBrick Classic is a popu...  
202  Bricks Breaker - Glow Balls is a addictive and...  
203  How to play\n- The ball flies to wherever you ...  
204  Fight brave soldiers from around the globe on ...

As you can see, there are two labelled columns: `name` (the name of the app) and `description` (a textual description). The code in the next cell shows how to access fields from the description column.

In [5]:
df['description'][200:205]

200    Introducing the best Brick Breaker game that e...
201    Classic Brick Game!\n\nBrick Classic is a popu...
202    Bricks Breaker - Glow Balls is a addictive and...
203    How to play\n- The ball flies to wherever you ...
204    Fight brave soldiers from around the globe on ...
Name: description, dtype: object

## Problem 1: Preprocessing

Your first task is to implement a preprocessor for your search engine. In the vector space model, *preprocessing* refers to any transformation applied to a text before vectorisation. Here you can restrict yourself to a simple type of preprocessing: tokenisation, stop word removal, and lemmatisation.

To implement your preprocessor, you can use [spaCy](https://spacy.io). Make sure to read the [Linguistic annotations](https://spacy.io/usage/spacy-101#annotations) section of the spaCy&nbsp;101; that section contains all the information you need for this problem (and more).

Implement your preprocessor by completing the skeleton code in the next cell, adding additional code as you deem necessary.

In [6]:

def preprocess(text):
    # TODO: Replace the next line with your own code.
    doc = nlp(text)
    list = []
    for token in doc:
        if not token.is_stop and token.is_alpha:
            list.append(token.lemma_)
    return list

Your implementation should conform to the following specification:

<strong>preprocess</strong> (<em>text</em>)

> Preprocesses the given text by tokenising it, removing any stop words, replacing each remaining token with its lemma (base form), and discarding all lemmas that contain non-alphabetical characters. Returns the list of remaining lemmas (represented as strings).

To speed up the preprocessing, you should disable loading those spaCy components you do not need, such as the parser, and named entity recogniser. See [here](https://spacy.io/usage/processing-pipelines#disabling) for more information about this.

### 🤞 Test your code

Test your implementation by running the following cell:

In [7]:
preprocess('Apple is looking at buying U.K. startup for $1 billion')

['Apple', 'look', 'buy', 'startup', 'billion']

This should give the following output:


## Problem 2: Vectorising

Your next task is to vectorise the data – and more specifically, to map each app description to a tf–idf vector. For this you can use the [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) class from [scikit-learn](https://scikit-learn.org/stable/). Make sure to specify your preprocessor from the previous problem as the `tokenizer` &ndash; not the `preprocessor`! &ndash; for the vectoriser. (In scikit-learn terminology, the `preprocessor` handles string-level preprocessing.)

In [8]:
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(tokenizer=preprocess)
X = vectorizer.fit_transform(df['description'])

### 🤞 Test your code

Test your implementation by running the following cell:

In [9]:
X.shape

(1614, 21396)

This should show the dimensions of the matrix `X` to be around 1614 × 21389. (The number of rows should be exactly 1614. The number of columns may differ from that given here depending on the version of spaCy and the version of the language model used, as well as the pre-processing.)

## Problem 3: Retrieving

To complete the search engine, your last task is to write a function that returns the most relevant app descriptions for a given query. An easy way to solve this task is to use scikit-learn&rsquo;s [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) class. That class implements unsupervised nearest neighbours learning and allows you to easily find a predefined number of app descriptions whose vector representations are closest to the query vector.

In [10]:
from sklearn.neighbors import NearestNeighbors
def search(query):
    # TODO: Replace the next line with your own code.
    neigh = NearestNeighbors(n_neighbors=10)
    neigh.fit(X)
    tquery = vectorizer.transform([query])
    results = neigh.kneighbors(tquery, return_distance=False)
    results2 = []
    for answer in results :
        results2.append(df['name'][answer])
    return results2

Your implementation should conform to the following specification:

<strong>search</strong> (<em>query</em>)

> Returns the 10 app descriptions most similar (in terms of cosine similarity) to the given query as a Pandas DataFrame.

### 🤞 Test your code

Test your implementation by running the following cell:

In [11]:
search('dodge trains')

[1301                         Subway Surfers
 1428                  Train Conductor World
 1394                             Tiny Rails
 1300                 Subway Princess Runner
 998          No Humanity - The Hardest Game
 1429    Train for Animals - BabyMagica free
 1168                                   Rush
 786                      LEGO® DUPLO® Train
 1465        Virus War - Space Shooting Game
 1153                              Road Riot
 Name: name, dtype: object]

The top hit in the list should be *Subway Surfers*.

## Problem 4: Finding terms with low/high idf

Recall that the inverse document frequency (idf) of a term is the lower, the more documents from a given collection the term appears in. To get a better understanding for this concept, your next task is to write code to find out which terms from the app descriptions have the lowest/highest idf.

Start by sorting the terms in increasing order of idf, breaking ties by falling back on alphabetic order.

In [12]:
# TODO: Replace the next line with your own code.
from collections import Counter 
import collections
import itertools
terms = []
Dict = {}
titles = vectorizer.get_feature_names_out()
testidf = vectorizer.idf_
index = 0
for title in titles :
    Dict[testidf[index]] = title
    index += 1

oDict = collections.OrderedDict(sorted(Dict.items()))
terms = list(oDict.values())


Now, print the 10 terms with the lowest/highest idf. How do you explain the results?

In [13]:
print(terms[:10])
print(terms[-10:][::-1])

['game', 'play', 'feature', 'free', 'new', 'world', 'time', 'app', 'fun', 'use']
['flye', '후', '확인', '저장', 'что', 'чтобы', '中文繁體', 'ไทย', '한국어', '日本語']


## Problem 5: Keyword extraction

We often want to extract salient keywords from a document. A simple method is to pick the $k$ terms with the highest tf–idf value. Your last task in this lab is to implement this method. More specifically, we ask you to implement a function `keywords` that extracts keywords from a text.

In [40]:
import operator
def keywords(text, n=10):
    # TODO: Replace the next line with your own code.
    tfidf = {}
    index = 0
        
    
    matrix = vectorizer.transform([text])
    temptext = preprocess(text)
    tempm = matrix.toarray()

    titles = vectorizer.get_feature_names_out()

    index = 0
    for title in titles:
        tfidf[title] = tempm[0][index]
        index += 1
    ordict = dict(sorted(tfidf.items(), key=lambda item: item[1]))    
    
    result = list(ordict.keys())
    return result[-n:][::-1]

Your implementation should conform to the following specification:

<strong>keywords</strong> (<em>text</em>, <em>n</em> = 10)

> Returns a list with the $n$ (default value: 10) most salient keywords from the specified text, as measured by their tf–idf value relative to the collection of app descriptions.

### 🤞 Test your code

Test your implementation by running the following cell:

In [41]:
print(keywords(df['description'][1428]))

['train', 'railway', 'railroad', 'rail', 'chaos', 'crash', 'tram', 'timetable', 're', 'railyard']


This should give output similar to the following:

The exact output may differ slightly depending on the strategy used to break ties.

## Reflection questions

The following reflection questions will help you prepare for the diagnostic test. Answer each of them in the form of a short text and put your answers in the cell below. You will get feedback on your answers from your lab assistant.

**RQ 1.1:** Why do we remove common stop words and lemmatise the text? Can you give an example of a scenario where, in addition to common stopwords, there are also *application-specific* stop words?

**RQ 1.2:** In Problem&nbsp;2, what do the dimensions of the matrix `X` correspond to? This matrix gives rise to different types of *representations*. Explain what these representations are. What information from the data is preserved, what information is lost in these representations?

**RQ 1.3:** What does it mean that a term has a high/low idf value? Based on this, how can we use idf to automatically identify stop words? Why do you think is idf not used as a term weighting on its own, but always in connection with term frequency (tf–idf)?

1.1
We remove common stop words since they don't add any information about the topic of the document for example 'the' will be used in almost every document and doesn't help us to extract document specific information. We lemmatise the words since different forms of different words usually have the same semantic meaning and differentiating between different forms therefore doesn't give us any additional information. An example of an application-specific stop word could be 'game' if we're looking at descriptions of games since almost every description will contain the word and thus it doesn't add information when analyzing different game descriptions.
1.2
The rows are the number of documents and the columns represent the amount of terms(words) in the dataset. The representations are the tf-idfs of all of the terms which is the term frequency multiplied with the inverse document frequency. This gives all the terms a weight that is supposed to represent how important the term is to the document. When we do this transformation we lose the semantic meaning of the descriptions since we only have a number for how important a term is but not what it means.
1.3
If a term has a high idf it means that the term is not in that many of the documents and a low idf means that the term is in almost all of the documents. Since a low idf means that a term is in almost all of the documents we can assume that it doesn't contain additional information. So one could automatically create stopwords from terms with a idf under a certain threshold. If only idf was used to as term-weighting rare words would get very high weights even if they are not used frequently in the document. And if a term is very frequent in a few documents it would be weighted lower than an unusual term used one time in one document. This in combination with our intuitive belief of how humans write(using topic specific words frequently) makes us believe that idf only weighted terms would give poor information about the documents.

**Congratulations on finishing L1! 👍**